### ISSUE TITLE
Group IDs and gaps

### ISSUE THREAD

**Issue by vasilvv** [moq-wg/moq-transport#427]
As far as I understand, we currently allow group IDs to have gaps in them, with only requirement for them is that they are monotonically increasing.  This presents a problem with caching.  Assume I am a relay, and I receive a range request for `[10, 20)`; if the origin has never produced group 15, then even if I have the entirety of that range in cache, I would not know that, since 15 would be just missing, and the only thing I can do with that is to request 15 from the origin every time.

Possible solutions to this problem:
1. Add some mechanism to indicate that 15 is not there (e.g. by putting the number of previous group in the beginning of every group).
2. Require group numbers to be incrementing by one.

I am fan of approach number 2 since it's conceptually simpler, but I think we can make some version of 1 work.

**Comment by suhasHere** [comment]
I am not fan of approach 2 since it is limiting and applies to typical media applications only.

I think we need to understand that, the application knows how the groupId increment. Catalog should provide that information for the players.

Regarding if there is a gap due to object being dropped, that information is identified within the object itself.

Let's take an application example where the group numbers increment by 2. So the groups would be 10,12,14,16,18,20 in the above example.
Now consider that group 12 was  not sent by the publisher , then the relay's cache will be in this form
[10, 12 (not sent by the publisher), 14,16,18, 20]

Keeping relay unaware of application semantics is something we should allow. Regardless, the end applications know exactly how the groups behave and if there are geps (either due to broadcaster not producing or dropped for other reasons), things will be identified as such

**Comment by afrind** [comment]
@suhasHere: can you give a specific use case for how an application could use non-sequential group IDs for some benefit?

**Comment by vasilvv** [comment]
> Let's take an application example where the group numbers increment by 2. So the groups would be 10,12,14,16,18,20 in the above example.
Now consider that group 12 was not sent by the publisher , then the relay's cache will be in this form
[10, 12 (not sent by the publisher), 14,16,18, 20]

The problem I have here is not with 12, the problem here is that the relay cannot distinguish 11,13,15,17,19 being in a state "has not arrived yet" vs state "does not and will never exist", so it will be always forced to fetch those.

**Comment by suhasHere** [comment]
> > Let's take an application example where the group numbers increment by 2. So the groups would be 10,12,14,16,18,20 in the above example.
> > Now consider that group 12 was not sent by the publisher , then the relay's cache will be in this form
> > [10, 12 (not sent by the publisher), 14,16,18, 20]
> 
> The problem I have here is not with 12, the problem here is that the relay cannot distinguish 11,13,15,17,19 being in a state "has not arrived yet" vs state "does not and will never exist", so it will be always forced to fetch those.

The relay should not worry about the semantic structure . It makes it brittle with any reasoning.  Relay needs to fetch only those that have been marked as dropped. If it hasn't arrived, then it will evetually arrive or the upstream will mark it as dropped.

**Comment by vasilvv** [comment]
I'm confused.  How is the relay supposed to know that when being asked for range `[10, 20)`, it does not need to fetch 15?

**Comment by suhasHere** [comment]
> I'm confused. How is the relay supposed to know that when being asked for range `[10, 20)`, it does not need to fetch 15?

Say relay cache is empty,  it will ask upstream from range [10-20) .. The upstream publisher will send 10,12,16,18, 20 .. Relay responds with that.  OTOH if the upstream provides 10, 12,, 16, 18(dropped due TTL), 20 and if the operation is fetch , the relay will fetch 18 to fill in the gap and once 18 arrives , it will be able to send 18 with others.

**Comment by vasilvv** [comment]
Sure, let's assume there are no drops, the first fetch is successful, and it puts 10, 12, 14, 16, 18 in cache.

Now the second fetch arrives for the exact same range, `[10, 20)`.  It looks at its cache; it has objects 10, 12, 14, 16, 18, that it can return immediately, but it doesn't know if that's all objects that exist in that range or not (we know that because we know that the previous request was `[10, 20)`, but it is entirely possible that those objects ended up in cache since they were results of five different fetches), so it has to repeat the fetch for those.

**Comment by kixelated** [comment]
I strongly think we should have sequential groups.

Like Victor said, the group ID can be used to detect gaps at the moq-transport layer if it's sequential. I understand the desire to stuff application-specific metadata into fields, but it's a layer violation. Put application-specific stuff in the object payload (ex. timestamps) and put properties useful to relays in the moq-transport layer.

**Comment by suhasHere** [comment]
GroupIds are just sync points and adding anything more inside relay is not in the moq-transport's scope . OTOH objects have dependencies within a group and having relay maintain a state to find out gaps may be useful.

**Comment by wilaw** [comment]
I am also a supporter of incrementally increasing Group IDs. Having such IDs does not break the contract that these points are independent of one-another. It simply means that they occur in a sequence. And being part of a sequence is a _fundamental attribute_ of all MOQT groups, since their parent, a track, represents  a temporal flow of data. 

> Now the second fetch arrives for the exact same range, [10, 20). It looks at its cache; it has objects 10, 12, 14, 16, 18, that it can return immediately, but it doesn't know if that's all objects that exist in that range or not (we know that because we know that the previous request was [10, 20), but it is entirely possible that those objects ended up in cache since they were results of five different fetches), so it has to repeat the fetch for those.

This example from Victor illustrates the core problem. For anything other than a contiguous sequence of group numbers, a relay can never know if the Groups exist or not and it would always have to make a request upstream to check. It could keep track of prior subscriptions (for example [10..12], [14..16], [18..20]), however if any of those did not overlap then it would be forced to make a fetch for [12..14] and [16..18] in case groups 13 and 17 existed. That is a wasted and convoluted workflow.

I tried to think of a use-case where I would want non-incremental group numbers. One example would be to use epoch time as group number. So if I'm making variable length segments, and doing segment-per-group, I might number my groups:

`1712302548317, 1712302550625, 1712302551901, 1712302553412`

A player wanting to seek back in time would need a timeline track to tell it the precise group number it needed. That timeline track could equally map group numbers to epoch time, which would allow me to number my groups incrementally:

`1,2,3,4`

The meta-point here is Group IDs only define sequence and any other relationship between those sync points can be expressed via an application level timeline track and/or catalog.

**Comment by suhasHere** [comment]
> And being part of a sequence is a _fundamental attribute_ of all MOQT groups, since their parent, a track, represents a temporal flow of data.

temporal flow of data doesn't imply sequential nature of groups either.

**Comment by suhasHere** [comment]
> Now the second fetch arrives for the exact same range, `[10, 20)`. It looks at its cache; it has objects 10, 12, 14, 16, 18, that it can return immediately, but it doesn't know if that's all objects that exist in that range or not

Second fetch gets the exact same answer (10,12,14,16,18 since it falls in the range). If there were gaps due to groups being dropped on path or a source deciding to skip a group , those should be marked for relay to see if it needs to satisfy fetch request.

**Comment by kixelated** [comment]
@suhas How does the subscriber know the SUBSCRIBE/FETCH is done? #424 

We have a SUBSCRIBE_DONE message with end=20. The subscriber receives 10,12,14,16,18,20 but when does it stop waiting? Is group 19 still in transit? There's no drop notification for it since it doesn't exist.

**Comment by suhasHere** [comment]
> @Suhas How does the subscriber know the SUBSCRIBE/FETCH is done? #424
> 
> We have a SUBSCRIBE_DONE message with end=20. The subscriber receives 10,12,14,16,18,20 but when does it stop waiting? Is group 19 still in transit? There's no drop notification for it since it doesn't exist.

Subscriber application knows its catalog and deduce the group distribution and find out there will be no grpup 19.

**Comment by kixelated** [comment]
> > @Suhas How does the subscriber know the SUBSCRIBE/FETCH is done? #424
> > We have a SUBSCRIBE_DONE message with end=20. The subscriber receives 10,12,14,16,18,20 but when does it stop waiting? Is group 19 still in transit? There's no drop notification for it since it doesn't exist.
> 
> Subscriber application knows its catalog and deduce the group distribution and find out there will be no grpup 19.

1. The moq-transport library should tell the application that a subscription has terminated, and not the other way around. Especially since the moq-transport library can't release any state until this optional signal is provided by the application (which might not care about gaps).

2. If the application needs to use a catalog/timeline to detect gaps, then it can use that same catalog/timeline to carry the timestamp that is being shoved into the `group_id`. Or it can put it in the object payload.

**Comment by kixelated** [comment]
So we're talking about two different fields and I'm going to use disambiguating names:

`group_sequence`: Increases by 1 each group. Provides ordering and gap detection.
`group_epoch`: Increases by an application specific amount for each group (no units). Provides ordering. 

There's a world where both are separate properties of a group in the moq-transport layer. Maybe the relay could use the epoch (as a timestamp with units) to perform TTLs, or latency budgets, or fetches.

But I think we should punt timestamps and metadata in general to the application for now. I'm also not sure how this would work in practice anyway, since there's no equivalent `object_epoch` as those are sequential.

**Comment by wilaw** [comment]
> temporal flow of data doesn't imply sequential nature of groups either.

We define a track (our subscribable entity) as a "sequence of Groups". That is part of our object model. There is no part of a track which does not belong to a group. Therefore, a track consists of a sequence of groups. 

![Screenshot 2024-04-05 at 5 19 54 PM](https://github.com/moq-wg/moq-transport/assets/2762250/643cf072-0a06-4b2a-b789-4292d5aa5848)

This diagram shows three tracks. The first two tracks are valid, one with temporally contiguous and the other with temporally non-contiguous groups. The third track is not valid, as we don't allow a publisher to produce objects at the same time in the same track which belong to different groups.

**Comment by suhasHere** [comment]
> This diagram shows three tracks. The first two tracks are valid, one with temporally contiguous and the other with temporally non-contiguous groups. The third track is not valid, as we don't allow a publisher to produce objects at the same time in the same track which belong to different groups.

this looks like an example constructed incorrectly. None of my points above said overlapping groups Ids. All I am saying is group Ids increase monotically and need not necessary increase in the units of 1.

**Comment by suhasHere** [comment]
> 2\. If the application needs to use a catalog/timeline to detect gaps, then it can use that same catalog/timeline to carry the timestamp that is being shoved into the `group_id`. Or it can put it in the object payload.

I think you missed the point in my comments . No catalog is not used to define gaps. Gaps are signaled in the data plane

**Comment by suhasHere** [comment]
> 1. The moq-transport library should tell the application that a subscription has terminated, and not the other way around. Especially since the moq-transport library can't release any state until this optional signal is provided by the application (which might not care about gaps).

Exactly. I am not arguing either. All I am saying is, when a subscribe done is processed and reported to the application, the application ha enough information to know if there needs to be group 19 or not, from your example. MoQ transport library reports status of what it has seen and only the application can know more about group 19 was supposed to be even produced or not , since that info is in catalog. Please note , this is nothing to do with GAPs.

**Comment by kixelated** [comment]
> > 1. The moq-transport library should tell the application that a subscription has terminated, and not the other way around. Especially since the moq-transport library can't release any state until this optional signal is provided by the application (which might not care about gaps).
> 
> Exactly. I am not arguing either. All I am saying is, when a subscribe done is processed and reported to the application, the application ha enough information to know if there needs to be group 19 or not, from your example. 

In the current state, since the moq-transport library cannot detect gaps, it has to immediately report any SUBSCRIBE_DONE to the application. OBJECTs may still be received during this state so the library does not drop state associated with the session. The application has to tell the library when the subscription is actually done so it can remove the lookup table entry.

With sequential groups and drop notifications, the moq-transport library can buffer the SUBSCRIBE_DONE until it has received and flushed all OBJECTs to the application. The application only learns the subscription is done after receiving the last object. This is how transport APIs generally work and how QUIC works itself.

**Comment by kixelated** [comment]
And just to draw inspiration from QUIC itself, it does not allow gaps in the Stream ID or Stream Offset. QUIC could have allowed the application to introduce gaps in either but it just complicates everything, as we're experiencing.

**Comment by suhasHere** [comment]
> And just to draw inspiration from QUIC itself, it does not allow gaps in the Stream ID or Stream Offset (roughly analogous to group_id and object_id). QUIC could have allowed the application to introduce gaps in either but it just complicates everything.

Not sure how to interepret this in the  context of this discussion. QUIC is transport layer protocol and MOQT is application protocol. Also  StreamIds and GroupIds are semantically different.

**Comment by afrind** [comment]
Chair Comment:

I think the WG *can* build the system to support either sequential or non-sequential groups.  

Several folks have put forward reasoning for why sequential groups provide benefit and/or why non-sequential groups add complexity.

@suhasHere as the lead proponent for non-sequential groups, can you state the use case and application benefits.  This will help us evaluate the tradeoffs.

**Comment by suhasHere** [comment]
> Several folks have put forward reasoning for why sequential groups provide benefit and/or why non-sequential groups add complexity.

@afrind  I am still not sure what are the complexities .. I have tried to answer all the questions raised. Can you please clarify ?

**Comment by vasilvv** [comment]
> Second fetch gets the exact same answer (10,12,14,16,18 since it falls in the range). If there were gaps due to groups being dropped on path or a source deciding to skip a group , those should be marked for relay to see if it needs to satisfy fetch request.

The problem is how to resolve this on the cache.  Right now, the cache, when it does not have any information about the state of object 15, cannot tell apart the scenario where 15 does not exist, vs the scenario where 15 is missing from the cache because the cache got filled by one client doing a `[10, 14]` request, and then another doing `[16, 20]`.

> For anything other than a contiguous sequence of group numbers, a relay can never know if the Groups exist or not and it would always have to make a request upstream to check. It could keep track of prior subscriptions (for example [10..12], [14..16], [18..20]), however if any of those did not overlap then it would be forced to make a fetch for [12..14] and [16..18] in case groups 13 and 17 existed. That is a wasted and convoluted workflow.

It is possible to work this around by putting a "does not exist" kind of entry into cache on FETCH: if I fetch `[10, 15]`, and get `10, 11, 15`, I can cache "`[12, 14]` does not exist" from my knowledge of the properties FETCH provides. Outside of being more complex, this also does not solve the entire problem, since our goal is to be able to cachefill from SUBSCRIBEs as well as FETCHes.  So non-sequential group IDs inherently makes caching less efficient.

**Comment by suhasHere** [comment]
> The problem is how to resolve this on the cache. Right now, the cache, when it does not have any information about the state of object 15, cannot tell apart the scenario where 15 does not exist, vs the scenario where 15 is missing from the cache because the cache got filled by one client doing a `[10, 14]` request, and then another doing `[16, 20]`.

Let's take the example of sequential groups for a similar use-case
Fetch 1: 10-14, cache is filled with 10, 11,12,13,14
Fetch 2: 18-20: cache is now filled with 10, 11,12,13,14, 18.19,20

now fetch 3 arrives with 10-20, Cache doesn't know the state of 15,16,17 . So as it needs to do,   it will have to ask upstream for `[15-18)`. 

When producer sends groups and objects, either as response to subscribe or fetch, the cache fill happens regardless. Let's say when subscription was made , an object was marked as dropped due to congestion. An eventual fetch request can end up replacing that object with proper object if it is still available upstream.

**Comment by vasilvv** [comment]
> Let's take the example of sequential groups for a similar use-case
Fetch 1: 10-14, cache is filled with 10, 11,12,13,14
Fetch 2: 18-20: cache is now filled with 10, 11,12,13,14, 18.19,20
>
> now fetch 3 arrives with 10-20, Cache doesn't know the state of 15,16,17 . So as it needs to do, it will have to ask upstream for [15-18). In the case where 15 was never produced for whatever reason, the publisher will need to mark it as such in the response regardless (like 15(not produced), 16, 17), as this will aid subsequent requests to not repeat the same request.

Sure, but in this case, fetch 3 will fill the cache, so if fetches 4, 5, 6 ask for `[15, 18)`, none of them will have to go to the origin.

Now, if we allow origin to never produce 16 and 17, and go from 15 directly to 18, the cache would never have any entry for 16 or 17, so _every single fetch_ for a range `[10, 20)` would have to go to the origin.

**Comment by vasilvv** [comment]
> In the case where 15 was never produced for whatever reason, the publisher will need to mark it as such in the response regardless (like 15(not produced), 16, 17), as this will aid subsequent requests to not repeat the same request.

Are you saying that the publisher is allowed to skip group IDs, but it has to explicitly mark groups in the middle as missing? This de-facto makes them sequential.

**Comment by suhasHere** [comment]
> > In the case where 15 was never produced for whatever reason, the publisher will need to mark it as such in the response regardless (like 15(not produced), 16, 17), as this will aid subsequent requests to not repeat the same request.
> 
> Are you saying that the publisher is allowed to skip group IDs, but it has to explicitly mark groups in the middle as missing? This de-facto makes them sequential.

This is the gap use-case that had a requirement  for players who want to know if a given group that is expected  is marked explicitly when source drops, thus allowing player to not wait forever.

**Comment by suhasHere** [comment]
> the only thing I can do with that is to request 15 from the origin every time.

Want to clarify that,  one needs to request for 15 only one time. Asking origin for non existent object ( never produced, not produced since origin decided to drop it or permanently gone from the origin) needs to be updated in the cache. Since a fetch can happen for object that fall in to  any of these cases and a relay that doesn't have that object will need to go back and find out the answer and it needs to do only once.

**Comment by kixelated** [comment]
> > In the case where 15 was never produced for whatever reason, the publisher will need to mark it as such in the response regardless (like 15(not produced), 16, 17), as this will aid subsequent requests to not repeat the same request.
> 
> Are you saying that the publisher is allowed to skip group IDs, but it has to explicitly mark groups in the middle as missing? This de-facto makes them sequential.

I want to second this. @suhasHere would allowing groups of size 0 work for you? 

The problem today is the ambiguity. A subscriber doesn't know if 15 is in flight or will never arrive. If the publisher explicitly tells the subscriber that 15 is empty (or dropped) then that solves things.

**Comment by suhasHere** [comment]
> I want to second this. @suhasHere would allowing groups of size 0 work for you?

yes it is totally possible and my point on " non existent object ( never produced, not produced since origin decided to drop it or permanently gone from the origin)" needs to be signaled regardless of how we design the groupIds.

**Comment by fluffy** [comment]
I don't get what the problem is here. The things the relays cache is the object. 

I am very concerned that any assumption on groups being sequential are being produced in order are not going to work with the way people want to use complex reference frames and layer in modern codecs like AV1.  Of course an given application like youtube might specify for it's usage of moq, everything was sequential. But the moq relays don't need to know this or assume this and keeping them flexible allows for all kinds of interesting media uses cases for media beyond video. 

I have helped write a relay, and I don't see what problem this is causing for relays.

**Comment by vasilvv** [comment]
> yes if there is a reason for a producer to drop entire groups and should be explicitly signaled.

Ah, my concern was that someone would try to have groups numbered 0, 1000, 2000, or use group ID as a timestamp, in which case you'd have to send hundreds of gap indicators.  If that is out of scope, simple "a group is missing" would work.

**Comment by suhasHere** [comment]
> > yes if there is a reason for a producer to drop entire groups and should be explicitly signaled.
> 
> Ah, my concern was that someone would try to have groups numbered 0, 1000, 2000, or use group ID as a timestamp, in which case you'd have to send hundreds of gap indicators. If that is out of scope, simple "a group is missing" would work.

That's my thinking too.  PR #429  provides aspects of that for the cases when the publisher explicitly drops a group IIRC.
